Support parsing variadic return schema

houseroad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23208 expose parse_schema and __eq__ function to python and add round trip tests
* #23207 Expose an API to iterate all the registered operators
* #23206 Align AliasInfo's operator<< with FunctionSchema
* #23205 Support IntList in Dict's shalloEquals
* #23204 Support variadic returns in Schema's operator<<
* #23203 Align the operator<< for Argument with FunctionSchema parser
* #23202 Fix the operator== for Argument
* #23201 Align String str() with the format in FunctionSchema
* #23200 Align Dict str() with the format in FunctionSchema
* **#23199 Support parsing variadic return schema**

all cases should be prim ops, but let's support it. it will expect variadic return schema to be prim::PythonOp(...) -> ...

Differential Revision: [D16431635](https://our.internmc.facebook.com/intern/diff/D16431635/)

houseroad committed:
Support parsing variadic return schema

all cases should be prim ops, but let's support it. it will expect variadic return schema to be prim::PythonOp(...) -> ...

Differential Revision: [D16431635](https://our.internmc.facebook.com/intern/diff/D16431635/)

houseroad committed:
Update on "Support parsing variadic return schema"


all cases should be prim ops, but let's support it. it will expect variadic return schema to be prim::PythonOp(...) -> ...

Differential Revision: [D16431635](https://our.internmc.facebook.com/intern/diff/D16431635/)

dzhulgakov commented on torch/csrc/jit/script/function_schema_parser.cpp:
```diff
@@ -58,7 +59,9 @@ struct SchemaParser {
     idx = 0;
     L.expect(TK_ARROW);
-    if (L.cur().kind == '(') {
+    if (L.nextIf(TK_DOTS)) {
```
`... foo()` does look a bit weird, shall we use `(...) foo` instead?

houseroad replied:
I would support both then :-)

dzhulgakov commented on torch/csrc/jit/script/function_schema_parser.cpp:
```diff
@@ -58,7 +59,9 @@ struct SchemaParser {
+      is_varret = true;
+    } else if (L.cur().kind == '(') {
       parseList('(', ',', ')', [&] {
         returns.push_back(
```
for completeness, we should support parsing `...` here too (just like above for the input args)

houseroad committed:
Update on "Support parsing variadic return schema"


all cases should be prim ops, but let's support it. it will expect variadic return schema to be prim::PythonOp(...) -> ...

Differential Revision: [D16431635](https://our.internmc.facebook.com/intern/diff/D16431635/)
